src/core/reaction_methods/ReactionAlgorithm.cpp

```diff
@@ -0,0 +1,651 @@
+/*
+ * Copyright (C) 2010-2019 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.hpp"
+
+#include "reaction_methods/ReactionAlgorithm.hpp"
+
+#include "energy.hpp"
+#include "grid.hpp"
+#include "partCfg_global.hpp"
+#include "particle_data.hpp"
+#include "statistics.hpp"
+
+#include <utils/Vector.hpp>
+#include <utils/constants.hpp>
+#include <utils/contains.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <iterator>
+#include <limits>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace ReactionMethods {
+
+/**
+ * Performs a randomly selected reaction in the reaction ensemble
+ */
+int ReactionAlgorithm::do_reaction(int reaction_steps) {
+  for (int i = 0; i < reaction_steps; i++) {
+    int reaction_id = i_random(static_cast<int>(reactions.size()));
+    generic_oneway_reaction(reaction_id);
+  }
+  return 0;
+}
+
+/**
+ * Adds a reaction to the reaction system
+ */
+void ReactionAlgorithm::add_reaction(
+    double gamma, const std::vector<int> &reactant_types,
+    const std::vector<int> &reactant_coefficients,
+    const std::vector<int> &product_types,
+    const std::vector<int> &product_coefficients) {
+  SingleReaction new_reaction(gamma, reactant_types, reactant_coefficients,
+                              product_types, product_coefficients);
+
+  // make ESPResSo count the particle numbers which take part in the reactions
+  for (int reactant_type : new_reaction.reactant_types)
+    init_type_map(reactant_type);
+  for (int product_type : new_reaction.product_types)
+    init_type_map(product_type);
+
+  init_type_map(non_interacting_type);
+
+  reactions.push_back(new_reaction);
+}
+
+/**
+ * Checks whether all necessary variables for the reaction ensemble have been
+ * set.
+ */
+void ReactionAlgorithm::check_reaction_ensemble() const {
```

Comment:
This function could be renamed to check_reaction_method()

Reply:
done